R/tbl_split.R

```diff
@@ -1,66 +1,263 @@
+#'     rows = variable == "grade" & row_type == "label"
+#'   ) |>
+#'   modify_spanning_header(
+#'     c(stat_1, stat_2) ~ "**TRT**"
+#'   ) |>
+#'   modify_abbreviation("I = 1, II = 2, III = 3") |>
+#'   modify_caption("_Some caption_") |>
+#'   modify_footnote_spanning_header(
+#'     footnote = "Treatment",
+#'     columns = c(stat_1)
+#'   ) |>
+#'   modify_source_note("Some source note!")
+#'
+#' # lets split it!
+#' tbl_list <- tbl_fc |>
+#'   tbl_split_by_rows(variables = c(marker, grade), footnotes = "last", caption = "first")
+#'
+#' @name tbl_split_by
 NULL
 
 #' @export
-#' @rdname tbl_split
-tbl_split <- function(x, ...) {
-  UseMethod("tbl_split")
+#' @rdname tbl_split_by
+tbl_split_by_rows <- function(x, variables, footnotes = "each", caption = "each") {
+  set_cli_abort_call()
+
+  # list map -------------------------------------------------------------------
+  if (inherits(x, "list")) {
+    tbl_list <- map(
+      .x = x,
+      .f = tbl_split_by_rows,
+      variables = {{ variables }}
+    )
+  } else {
+    # check inputs ---------------------------------------------------------------
+    check_class(x, "gtsummary")
+
+    # process inputs -------------------------------------------------------------
+    cards::process_selectors(
+      data = scope_table_body(x$table_body),
+      variables = {{ variables }}
+    )
+    # adding last variable
+    variables <- variables |> union(dplyr::last(x$table_body$variable))
+
+    footnotes <- arg_match(footnotes, values = c("each", "first", "last"))
+    # if (missing(footnotes)) {
+    #   footnotes <- get_theme_element("tbl_summary-arg:footnotes", default = translate_string(footnotes)) # styler: off
+    # }
+    check_string(footnotes)
+
+    # caption check
+    caption <- arg_match(caption, values = c("each", "first"))
+    check_string(caption)
+
+    # merging split points -------------------------------------------------------
+    # convert list of table_body into list of gtsummary objects
+    tbl_list <- x$table_body %>%
+      dplyr::left_join(
+        dplyr::tibble(variable = variables, ..group.. = variables),
+        by = "variable"
+      ) |>
+      tidyr::fill("..group..", .direction = "up") |>
+      tidyr::nest(data = -"..group..") |>
+      dplyr::pull("data") |>
+      map(
+        ~ list(.) |>
+          set_names("table_body") |>
+          c(utils::modifyList(x, val = list(table_body = NULL))) %>% # add the other parts of the gtsummary table
+          `class<-`(class(x)) # add original class from `x`
+      )
+
+    # caption/footnotes handling -----------------------------------------------
+    if (length(tbl_list) > 1) {
+      tbl_list <- .modify_footnotes_caption(tbl_list, footnotes, caption)
+    }
+  }
+
+  # return list of tbls --------------------------------------------------------
+  # assign class (can't assign gtsummary because of print definitions)
+  tbl_list |>
+    structure(class = c("tbl_split", "tbl_split_by_rows", "list"))
```

Comment:
The most general classes go at the end, so this should be `c("tbl_split_by_rows", "tbl_split", "list")`